mmverify.py

```diff
@@ -551,12 +551,18 @@ def treat_compressed_proof(
         """Return the proof stack once the given compressed proof for an
         assertion with the given $f and $e-hypotheses has been processed.
         """
+        active_hypotheses = {label for frame in self.fs for labels in (frame.f_labels, frame.e_labels) for label in labels.values()}
         # Preprocessing and building the lists of proof_ints and labels
         flabels = [self.fs.lookup_f(v) for _, v in f_hyps]
         elabels = [self.fs.lookup_e(s) for s in e_hyps]
         plabels = flabels + elabels  # labels of implicit hypotheses
         idx_bloc = proof.index(')')  # index of end of label bloc
-        plabels += proof[1:idx_bloc]  # labels which will be referenced later
+        label_list = proof[1:idx_bloc]
+        def authorized(label):
+            return label in active_hypotheses or self.labels.get(label)[0] in {'$a', '$p'}
```

Comment:
I suppose `self.labels` only contains labels _prior_ to the current theorem?
This way it is forbidden to reference later theorems.